cpp/src/fil/fil.cu

```diff
@@ -322,24 +346,32 @@ struct dense_forest : forest {
   virtual void infer(predict_params params, cudaStream_t stream) override {
     dense_storage forest(nodes_, num_trees_,
                          algo_ == algo_t::NAIVE ? tree_num_nodes(depth_) : 1,
-                         algo_ == algo_t::NAIVE ? 1 : num_trees_);
+                         algo_ == algo_t::NAIVE ? 1 : num_trees_, vector_leaf_);
     fil::infer(forest, params, stream);
   }
 
   virtual void free(const raft::handle_t& h) override {
     int num_nodes = forest_num_nodes(num_trees_, depth_);
     h.get_device_allocator()->deallocate(nodes_, sizeof(dense_node) * num_nodes,
                                          h.get_stream());
+    if (vector_leaf_len_) {
+      h.get_device_allocator()->deallocate(
```

Comment:
I can see at least one case in raft where it avoids deallocating a null pointer: https://github.com/rapidsai/raft/blob/fedab7674c05989da9155aed3a11e241113e3397/cpp/include/raft/mr/buffer_base.hpp#L109
It should be a no-op but I guess it's implementation dependent, so this would be safer.